### PR TITLE
Fixes jupyter/jupyterlab#943: added an explicit cell list to the console

### PR DESCRIPTION
**Pull request by akosyakov**

### CODE REVIEW

**Comment by blink1073** [comment]
LGTM, but needs a rebase after #953.

File: src/console/content.ts [hunk 1]
```diff
@@ -533,6 +538,24 @@ class ConsoleContent extends Widget {
     this.prompt.mimetype = this._mimetype;
   }
 
+  protected addContentCell(cell: CodeCellWidget) {
+    this._content.addWidget(cell);
+    const prompt = this.prompt;
+    const index = prompt ? this._cells.indexOf(prompt.model) : -1;
```
Comment (afshin) [inline src/console/content.ts]: I don't understand the logic here. It seems as though `prompt` will always be defined.

Comment (akosyakov) [inline src/console/content.ts]: Do you mean it is always safe to call `this._cells.indexOf` without a check?

Comment (afshin) [inline src/console/content.ts]: My confusion stems from this: perhaps a cell model should be added to `_cells` only when it is ready to go into `_content` and not when the prompt is created. That way an `indexOf` is unnecessary, it's always just added to the end of the list. I would propose adding to `_cells` here: https://github.com/R-Brain/jupyterlab/blob/bfb5b04f7b7cfd73d90f7fb6525cfca6b570473c/src/console/content.ts#L441

Comment (akosyakov) [inline src/console/content.ts]: I see, makes sense, please have a look again

**Comment by afshin** [comment]
@akosyakov Oh yeah! One other thing that is important. Please make sure the `clear()` method is updated to remove the cell models as well when the console content is emptied.

**Comment by akosyakov** [comment]
@afshin done

I actually wonder about memory management in JS. I've learned that modern browsers have GC similar to Java, so if you even have circular dependencies they are able to collect unreachable objects. So one don't really write code to reset references to null, but tweak code a bit when there is an actual issue, like to make sure that some critical resources are not hold. I wonder whether aggressive disposing does any help.

I've started wondering after playing with `strictNullCheck` in typescript 2. If you enable it then, for example, `ConsoleContent._input` should be of a union type with Null type to work around reseting it to null in `dispose` method, but it is confusing, since it should not be null but initialized in the constructor, and lead to null checking of `_input` on each access.

**Comment by akosyakov** [comment]
I think it is failed because of some flaky test

**Comment by afshin** [comment]
I restarted the test.

I don't think we've have the non-nullable types discussion yet, but I think mostly that seems like a good direction to go. You may want to raise the issue at the JupyterLab weekly dev meeting.

Please check out my comment above about the `clear()` method, though, because right now, there is an out-of-sync issue if somebody clears the cells in the console.

**Comment by akosyakov** [comment]
@afshin a model should be removed now when a cell is disposed: https://github.com/jupyter/jupyterlab/pull/951/files#diff-024bf3dea7b0061f4729d21543bafdbdR544

**Comment by ellisonbg** [comment]
@akosyakov looks like there are a few more things to fix here before we can merge. We will watch this to see when it is ready to go

**Comment by afshin** [comment]
Thanks, @akosyakov! I forked your branch and picked up where you left off with the rebase and requested changes, so I'm closing this PR as your work is represented here: https://github.com/jupyterlab/jupyterlab/pull/1174

**Pull request closed without merging**